### ISSUE TITLE
Problem with Central Server listing limit of 200 servers

### ISSUE THREAD

**Issue by gilgongo** [corrados/jamulus#50]
The problem with the 200 server limit is that somebody sets up a public server, but neither they nor their bandmates can see it listed because the central server is full at the time. And with so many servers being registered and few leaving, they may never see it listed.

But with no indication of the real problem, they then think they have a technical issue with their setup,. So they play with port forwarding and firewalls, post on the forums, etc. until they eventually conclude they need to set up a private server. At which point they they might get other problems, post on the forums about port forwarding, ISP issues, etc.

Is it worth addressing this issue, and if so, how (other than raising the limit)?

**Comment by pljones** [comment]
Any server with a red ping distance is probably not worth trying.  If your listing is full with no red servers showing, then maybe the list needs to be longer.

**Comment by gilgongo** [comment]
Sorry - I should have explained better. 

The problem I think exists is that somebody sets up a public server, but neither they nor their bandmates can see it listed by the central server because the central server is full at the time. And with so many servers being registered and few leaving, they may never see it listed.

But with no indication of the real problem, they then think they have a technical issue with their setup, play with port forwarding and firewalls, post on the forums, etc. until they eventually conclude they need to set up a private server. At which point they they might get other problems, post on the forums about port forwarding, ISP issues, etc.

**Comment by corrados** [comment]
We should not increase the limit. The list is already sluggish. And each time a client opens the list, permanently ping messages, client list messages are transmitted. So the network load increases with the length of the list.

**Comment by gilgongo** [comment]
Indeed, that is why I say "other than raising the limit".

But assuming we think the problem as I describe it **is** a problem (and I don't know for sure if it is that's why I'm asking) - how might we address it? 

One short term way would be to have a warning message to allow server operators to know that their server is not registered. At least that way they can know what the problem is.

Longer term, we could encourage others to set up alternative central servers, then modify the client and server behaviour in some way such that Jamulus can scale. Servers might have a choice of central servers to register with, while clients might show lists from multiple central servers, for example. This of course would be an "architectural" change - a bit like moving from using hosts files to DNS servers was with TCP/IP.

I'm sure there are many other ways to approach it though.

**Comment by corrados** [comment]
Ok, I think we have to address this issue. One idea would be to have a Central Server for each continent. As a start, e.g., Europe, America, Asia. These Central Servers could all run at the server-hardware in Frankfurt. That would be just additional Jamulus instances running in parallel. I already do this for the 46 samples frame size version, see this post: https://sourceforge.net/p/llcon/discussion/533517/thread/bdb3fe2195/#1dcf. This server just uses the port number 22064 instead.

In the Jamulus settings dialog we have to possibility to set the Central Server address and there is a default check box. Instead of the check box we could introduce a combo box with: {Default Europe, Default America, Default Asia, Manual Entry}. With the Qt-locale, as we do for the Musician Profile dialog, we can derive from the OS settings in which country you are. Based on that country we could choose the correct default Central Server for the correct region.

**Comment by corrados** [comment]
Ok, first step is done. I setup a new Central Server for North America (just one additional right now). It has a special port number assigned. To use that new Central Server, setup your server to use:

jamulus.fischvolk.de:22224

Also in the Jamulus client you have to set this address as the Central Server Address field.
Next step will be to include the new port in the client/server settings dialog. I will take a look at this now.

**Comment by gilgongo** [comment]
What do you think about the idea of introducing a status indication of some kind so that users know whether their server is or not currently registered?

**Comment by corrados** [comment]
Please create a separate Issue for that.

**Comment by pljones** [comment]
Ideally an incoming request from a client could have its IP geolocated and either have the request served or, where there's a "better" known central server, respond telling the client to ask there instead.  That would gradually move clients over to the other servers.

**Comment by corrados** [comment]
We'll start with the solution I proposed above. Most of the coding is already done this morning. I will apply a lot of new code to the Git repo today. Then I need volunteers to test the new implementation.

**Comment by pljones** [comment]
I can spin up a "pretend" server to connect to any "regional central" server when you're ready.

**Comment by corrados** [comment]
Ok, I added new code to Git: https://github.com/corrados/jamulus/commit/82f5ad18f4feeb34aa2618105e957268d78af068
This code is almost untested. Any feedback appreciated.

**Comment by gilgongo** [comment]
As an aside on this, here's [an interesting take on the situation ](https://sourceforge.net/p/llcon/discussion/server/thread/dfff628fd8/#7fc1) which doesn't necessarily go against the current solution: drop the server if it has been idle for a set time.

As I say in my reply to him, while it breaks the current Jamulus model, expiring empty servers from the list might be closer to the way people expect to behave for jamming as opposed to rehearsing. Those who wish to jam with friends or strangers can "declare" this by starting their servers in public mode, and connecting to the Central Server for the duration of the jam. Every server would then have a minimum of one person in it, who by definition would be wanting to play with anyone who arrived.

Meanwhile, the those who wish to rehearse with bandmates can do so on private servers. 

Seems worth thinking about at least.

**Comment by corrados** [comment]
3.4.7 is released. Let‘s see if it relaxes the situation. If not, we can reopen this issue again.

**Comment by vocobox** [comment]
@corrados said
> We should not increase the limit. The list is already sluggish. And each time a client opens the list, permanently ping messages, client list messages are transmitted. So the network load increases with the length of the list.

Do you think the ping accuracy depends on the number of ping request sent in parallel? Would the ping time drop if there were 2000 servers in the list?

**Comment by corrados** [comment]
See this Issue: https://github.com/corrados/jamulus/issues/49
